src/main/java/com/eternalcode/discordapp/experience/listener/ExperienceVoiceListener.java

```diff
@@ -0,0 +1,81 @@
+package com.eternalcode.discordapp.experience.listener;
+
+import com.eternalcode.discordapp.data.YamlFilesManager;
+import com.eternalcode.discordapp.experience.ExperienceConfig;
+import com.eternalcode.discordapp.experience.ExperienceService;
+import com.eternalcode.discordapp.experience.data.UsersVoiceActivityData;
+import net.dv8tion.jda.api.events.guild.voice.GuildVoiceUpdateEvent;
+import net.dv8tion.jda.api.hooks.ListenerAdapter;
+
+import java.time.Instant;
+import java.time.LocalTime;
+import java.time.ZoneId;
+
+public class ExperienceVoiceListener extends ListenerAdapter {
+
+    private final ExperienceConfig experienceConfig;
+
+    private final UsersVoiceActivityData usersVoiceActivityData;
+    private final YamlFilesManager yamlFilesManager;
+    private final ExperienceService experienceService;
+
+    public ExperienceVoiceListener(ExperienceConfig experienceConfig, UsersVoiceActivityData usersVoiceActivityData, YamlFilesManager yamlFilesManager, ExperienceService experienceService) {
+        this.experienceConfig = experienceConfig;
+        this.usersVoiceActivityData = usersVoiceActivityData;
+        this.yamlFilesManager = yamlFilesManager;
+        this.experienceService = experienceService;
+    }
+
+    @Override
+    public void onGuildVoiceUpdate(GuildVoiceUpdateEvent event) {
+        try {
+            if (event.getMember().getUser().isBot()) {
+                return;
+            }
+
+            this.leaveVoiceChannel(event);
+            this.joinVoiceChannel(event);
+
+            this.yamlFilesManager.save(this.usersVoiceActivityData);
+        }
+        catch (Exception exception) {
+            throw new RuntimeException(exception);
+        }
+    }
+
+    private void leaveVoiceChannel(GuildVoiceUpdateEvent event) {
+        if (event.getChannelLeft() == null) {
+           return;
+        }
+
+        long userId = event.getMember().getIdLong();
+        this.usersVoiceActivityData.usersOnVoiceChannel.remove(event.getMember().getIdLong());
+        this.experienceService.addPoints(userId, this.calculatePoints(event));
+    }
+
+    private void joinVoiceChannel(GuildVoiceUpdateEvent event) {
+        long userId = event.getMember().getIdLong();
+        if (this.usersVoiceActivityData.usersOnVoiceChannel.containsKey(userId)) {
+            return;
+        }
+
+        if (event.getChannelJoined() == null) {
+            return;
+        }
+
+        this.usersVoiceActivityData.usersOnVoiceChannel.put(userId, Instant.now().toEpochMilli());
+    }
+
+    private double calculatePoints(GuildVoiceUpdateEvent event) {
+        long userId = event.getMember().getIdLong();
+        
+        long timeSpentOnChannel = Instant.now().getEpochSecond() - this.usersVoiceActivityData.usersOnVoiceChannel.get(userId);
+        Instant instant = Instant.ofEpochSecond(timeSpentOnChannel);
+        LocalTime localTime = LocalTime.ofInstant(instant, ZoneId.systemDefault());
```

Comment:
Po co local time jest używane?

Reply:
By wyliczyć ile minut użytkownik spędził na kanale voice

Reply:
Instant i LocalTime to są punkty w czasie (a **nie** czas trwania) np. 2001 rok 1 września

Duration to czas trwania np. 2h 20min

Między 2 momentami w czasie (Instant) możesz obliczyć czas trwania (Duration) 

np. 
Duration duration = Duration.between(joinMoment, Instant.now());

 I nie używaj longów.